cmd/promtool/backfill.go

```diff
@@ -0,0 +1,176 @@
+func NewParser(r io.Reader) textparse.Parser {
+	return &OpenMetricsParser{s: bufio.NewScanner(r)}
+}
+
+// Next advances the parser to the next sample. It returns io.EOF if no
+// more samples were read.
+func (p *OpenMetricsParser) Next() (textparse.Entry, error) {
+	for p.s.Scan() {
+		line := p.s.Bytes()
+		line = append(line, '\n')
+		p.Parser = textparse.New(line, string(expfmt.FmtOpenMetrics))
+		if et, err := p.Parser.Next(); err != io.EOF {
+			return et, err
+		}
+	}
+	if err := p.s.Err(); err != nil {
+		return 0, err
+	}
+	return 0, io.EOF
+}
+
+func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
+	var maxt, mint int64 = math.MinInt64, math.MaxInt64
+	for {
+		entry, err := p.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, 0, errors.Wrap(err, "next")
+		}
+		if entry != textparse.EntrySeries {
+			continue
+		}
+		l := labels.Labels{}
+		p.Metric(&l)
+		_, ts, _ := p.Series()
+		if ts == nil {
+			return 0, 0, errors.Errorf("expected timestamp for series %v, got none", l.String())
+		}
+		if *ts > maxt {
+			maxt = *ts
+		}
+		if *ts < mint {
+			mint = *ts
+		}
+	}
+	if maxt == math.MinInt64 {
+		maxt = 0
+	}
+	if mint == math.MaxInt64 {
+		mint = 0
+	}
+	return maxt, mint, nil
+}
+
+func createBlocks(input *os.File, mint, maxt int64, outputDir string) error {
+	var offset int64 = 2 * time.Hour.Milliseconds()
+	mint = offset * (mint / offset)
+	for t := mint; t <= maxt; t = t + offset {
+		err := func() error {
+			errL := ListBlocks(outputDir, true)
+			if errL != nil {
+				return errors.Wrap(errL, "print blocks")
+			}
+			w, err := tsdb.NewBlockWriter(log.NewNopLogger(), outputDir, offset)
+			if err != nil {
+				return errors.Wrap(err, "block writer")
+			}
+			defer func() {
+				err = tsdb_errors.NewMulti(err, w.Close()).Err()
+			}()
+			ctx := context.Background()
+			app := w.Appender(ctx)
+			_, errReset := input.Seek(0, 0)
+			if errReset != nil {
+				return errors.Wrap(errReset, "seek file")
+			}
+			p := NewParser(input)
+			tsUpper := t + offset
+			var samplesCount int64
+			for {
+				e, errP := p.Next()
+				if errP == io.EOF {
+					break
+				}
+				if errP != nil {
+					return errors.Wrap(errP, "parse")
+				}
+				if e != textparse.EntrySeries {
+					continue
+				}
+				l := labels.Labels{}
+				p.Metric(&l)
+				_, ts, v := p.Series()
+				if ts == nil {
+					return errors.Errorf("expected timestamp for series %v, got none", l.String())
+				}
+				if *ts >= t && *ts < tsUpper {
+					samplesCount++
+					if _, err := app.Add(l, *ts, v); err != nil {
+						return errors.Wrap(err, "add sample")
+					}
+				}
+				if samplesCount >= 5000 {
+					err := app.Commit()
+					if err != nil {
+						return errors.Wrap(err, "commit")
+					}
+					app = w.Appender(ctx)
+					samplesCount = 0
+				}
+			}
+			if err := app.Commit(); err != nil {
+				return errors.Wrap(err, "commit")
+			}
+			_, errF := w.Flush(ctx)
+			if tsdb.SeriesCount != 0 && errF != nil {
```

Comment:
See below, you cannot just make this a top-level exported variable.

How about checking what error you got back instead?

`if errF.Error() == "no series appended, aborting" { `

Reply:
We cannot do this because this error is created when seriesCount is 0 in Flush. Instead, I made a new function in blockwriter that returns the series count.

Reply:
this i think would mean that throw error when seriescount = 0, but we do not want to throw error then. So, I made a function in blockwriter

Reply:
What I meant is to not return an error `if errF.Error() == "no series appended, aborting"` and to not throw an error if `errF == nil`, but return `errF` in all other cases.

So this line should be `if errF != nil && errF.Error() != "no series appended, aborting" {`

Then, if I'm not mistaken, you shouldn't need a method in the blockwriter.

Reply:
yes 👍